core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedQueryScheduler.java

```diff
@@ -847,12 +855,14 @@ public static DistributedStagesScheduler create(
         {
             DistributedStagesSchedulerStateMachine stateMachine = new DistributedStagesSchedulerStateMachine(queryStateMachine.getQueryId(), executor);
 
+            int hashPartitionCount = getEstimatedHashPartitionCount(stageManager, getMaxHashPartitionCount(queryStateMachine.getSession()));
             Map<PartitioningHandle, NodePartitionMap> partitioningCacheMap = new HashMap<>();
             Function<PartitioningHandle, NodePartitionMap> partitioningCache = partitioningHandle ->
                     partitioningCacheMap.computeIfAbsent(partitioningHandle, handle -> nodePartitioningManager.getNodePartitioningMap(
                             queryStateMachine.getSession(),
                             // TODO: support hash distributed writer scaling (https://github.com/trinodb/trino/issues/10791)
-                            handle.equals(SCALED_WRITER_HASH_DISTRIBUTION) ? FIXED_HASH_DISTRIBUTION : handle));
+                            handle.equals(SCALED_WRITER_HASH_DISTRIBUTION) ? FIXED_HASH_DISTRIBUTION : handle,
+                            hashPartitionCount));
```

Comment:
`PartitioningHandle` is a key for `partitioningCacheMap`. It seems we don't really need to have global `hashPartitionCount` for all stages.

I'm inclining that we could move `Optional<Int> hashPartitionCount` directly into `SystemPartitioningHandle` since we already have special handling of these AND `SystemPartitioningHandles` are (almost) the only usages of `hashPartitionCount`.

When we would make `hashPartitionCount` more generic (e.g. that involves connector provided `ConnectorPartitioningHandles`), then we could move it into `PartitioningHandle` (the lowest possible level)

It would also simplify `io.trino.sql.planner.NodePartitioningManager#getBucketCount` and `io.trino.sql.planner.NodePartitioningManager#getNodePartitioningMap(io.trino.Session, io.trino.sql.planner.PartitioningHandle)` since we wouldn't need an extra parameter.